Make the code easier to read with new features from Rust 1.65

jplatte opened the pull request:
It took a while, but don't be discouraged from pushing back on any of the individual changes or the whole PR if you don't find the code more readable than before.

The last commit in particular might not be a good idea, maybe we can factor out parts into a new function there?

Last remark: Unfortunately rustfmt doesn't yet format let-else code at all. Please complain if you find the formatting weird in some place.

poljar commented:
I have not looked at the code yet, but doesn't this need to bump the `rust-version` field in `Cargo.toml` files, which don't seem to be touched.

jplatte commented:
Yeah I had a commit for it locally, but failed to include it in this PR. Fixed.

Hywan reviewed (approved):
It looks fine to me. Thanks for taking care of this!

gnunicorn commented:
only read parts of it, but gotta say:

so yum
very style
such taste

like

jplatte commented:
No concerns about the last commit? 😄 

@poljar do you want to take a closer look before this is merged?

poljar commented:
> No concerns about the last commit? smile
> 
> @poljar do you want to take a closer look before this is merged?

Looking.

poljar commented on crates/matrix-sdk-crypto/src/gossiping/machine.rs:
```diff
@@ -315,76 +315,74 @@ impl GossipMachine {
-                }
-            }
-        } else {
+        let Some(device) = device else {
```
I think I can get behind the device `None` case early return, but for the rest I would prefer the whole `match self.forward_room_key(&session, &device, message_index).await` block to become a separate method.

Alternatively, just leaving out the second early return works for me as well.

jplatte replied:
I don't understand exactly what you mean. Can you elaborate, or just push a commit with your preferred solution?

poljar replied:
Did so in https://github.com/matrix-org/matrix-rust-sdk/pull/1185/commits/852d16d09224e040d1dcc4f98f28d9ded5ae0522.

poljar reviewed (changes requested):
Sorry, there are a few changes that I think are a downgrade. I like to avoid early returns, granted sometimes a bit too much. 

The reasoning behind that is that when we add an early return, we linearize a branched execution tree using jumps. The tree still exist, we need to reconstruct it in our head while we read through the particular piece of code. The tree has been hidden from us. If we don't have early returns, we can see the tree right in front of us.

Now not every jump is bad, that's why the `?` operator is neat. Some early returns in this patch series were quite nice improvements, thanks.

jplatte commented on crates/matrix-sdk-crypto/src/backups/mod.rs:
```diff
@@ -241,43 +241,43 @@ impl BackupMachine {
 
         if let Some(user_signatures) = signatures.get(self.account.user_id()) {
             for device_key_id in user_signatures.keys() {
-                if device_key_id.algorithm() == DeviceKeyAlgorithm::Ed25519 {
```
Does the `No need to check our own device here` comment apply regardless of the algorithm? I.e. should the `device_key_id.device_id() == self.account.device_id()` check be part of the new function or move to before the fn call?

poljar replied:
I would keep it as part of the algorithm, so we don't end up in a scenario where we accidentally do something with an unknown algorithm and attackers gain the ability to spoof algorithms trying to get us to mess up.

It does look like it would be fine either way but I prefer them to be completely separate codepaths.

jplatte commented:
Alright, I rebased, removed the labeled-block commit, factored out a new method in one case and reverted two changes from the early-return commit. Please take a look at the latest two commits.
